Ilya - Consumables: Single Update

ilyaflaks opened the pull request:
<img width="755" alt="Screenshot 2024-02-29 at 6 17 42 PM" src="https://github.com/OneCommunityGlobal/HGNRest/assets/105448137/ce62ea76-ad37-486a-8eeb-6df7fea2518b">

# Description
Adds functionality for the user to navigate to [/bmdashboard/consumables](http://localhost:3000/bmdashboard/consumables), select a consumable that needs to have it’s stock updated, press the pencil icon, enter how much of the consumable was used or wasted (in % or units) and press “Update Consumable”. The change should update the Consumables table right away. 

## Related PRS:
To test this frontend PR you need to checkout the [#780](https://github.com/OneCommunityGlobal/HGNRest/pull/780) backend PR.

## Main changes explained:
- New component UpdateConsumable created
- Real-time validation added to warn the user when the quantity entered exceeds current stock
- API calls handled through Redux actions. An action to fetch all consumables is dispatched after the call to update the table with the changes without having to reload the page

## How to test:
1. checkout **ilya_consumables_single_update** branch on the back end. Make sure you are using Node version 14.17.6. Run **npm install** and **npm run build && npm start** to run the backend locally
2. check into current branch on the frontend. Run commands **npm install** and **npm run start:local** to launch the front end application
3. log in as any user
4. go to [http://localhost:3000/bmdashboard](http://localhost:3000/bmdashboard) and log in again
5. go to [http://localhost:3000/bmdashboard/consumables](http://localhost:3000/bmdashboard/consumables), select any item on the table, press the “pencil” icon, enter some numbers next to “Quantity Used” and/or “Quantity Wasted”. Select "%" as units or leave the item's default units 
6. Press “Update Consumable”
7. A success toast message should appear if everything is acceptable or a warning message if not
8. From the Consumables table, press “View” in the “Updates” column and verify that the change you made is reflected. The latest update should be on the bottom

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/105448137/83417ee2-5ce8-47ae-84af-fbdf51776d13


## Note:
This change will update the “buildingInventoryItems” collection on the database. As an alternative to step 8 you can open Compass, search the collection by the criteria {"__t":"consumable_item”} and verify that updateRecord property has a new entry with today’s date.

nnamdixobi commented:
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/80783443/68a8980a-0245-4c3e-ab10-74cdffef81c8

Definitely works as expected and got no errors from the code either

Weiyao-Li commented:
Hi, I tried to test this pr. However, even though I successfully executed the commands in front end and backend, I cannot open the website. To make sure it is not the setup error from my end, I also checkout other pr, which works as expected.

<img width="1705" alt="image" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/96925606/563807ae-87d6-40b0-9acf-432a340a992d">
<img width="1340" alt="image" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/96925606/ad2b56bb-9f4e-4664-815b-bdb14613a5c4">
<img width="1543" alt="image" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/96925606/93d25085-fcdc-4258-b95e-797ec621b725">

DiegoSalas27 commented:
Hi Ilya,

I have submitted a video showcasing the functionality you've implemented. The form you created accept negative numbers and characters such as '-' or '+'. I do believe this is an error. In addition, I am pointing out some other concerns regarding the PR submissions and are outside of the scope of your PR


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/26840798/4fd67d2b-8d53-4ca8-90db-59cabf9fb149

dhairya-mehta commented:
Hey Ilya, 
the website doesn't seem to be loading

ilyaflaks commented:
@DiegoSalas27 First of all, thank you for reviewing my PR and catching this bug! I definitely overlooked some edge cases. 
My latest commit resolves the negative number issue. I’ve looked into several possibilities to address the “-“ or “+” in the middle of the input and did not find a suitable solution. What happens there is when the event is passed from the Input component to the handler, the number is already converted to a string. If the input is non-sensical, such as “100-110” it returns an empty string (“”). This is a behavior of the HTML input tag of type "number", not the Reactstrap component. 
Because of this behavior, I can’t restrict that specific case without also restricting an empty string entry, which is what happens when a user enters a number and then deletes it. Submitting the form with this input is not going to have harmful effects on the app because the value will be converted to 0. 
I added a validation for the case when an input is something like “100-110” and the other input (Quantity Wasted, for instance) is zero or “”. In this scenario the “Update” button will be disabled and nothing will happen. In the edge case when one entry is nonsensical and the other is valid, the nonsensical one will be ignored. Given a choice between bad UI and addressing a rare edge case I decided to choose in favor of the UI. 
If you know a better approach on the front end, please let me know. 
I also added an extra validation on the back end to send a 200 status when either number is negative. It should never happen if the front end is working correctly but having another level of protection is never a bad idea. 

As for the linting issue, I agree that it could be a little annoying having to scroll through a list of changes that are actually just semicolons and spaces. In theory, previous PR’s with these changes that were merged should have eliminated this problem. I’m not sure why it still exists. If you know the workaround, I suggest you bring it up to Jae, so that he can update the [document on using ESLint and Prettier](https://docs.google.com/document/d/1-zXEU1Go1RheI0lNSY9v8bnvFUjOkp7oJHDT1CxuowM/edit#heading=h.emft21z1gsuy). Right now there is no official policy regarding this. 

@Weiyao-Li and @dhairya-mehta You are right, there was a conflict between my branch and the current version of the “development” branch. I pulled the changes and committed again. Please retest at your convenience. Thank you

DiegoSalas27 commented:
> @DiegoSalas27 First of all, thank you for reviewing my PR and catching this bug! I definitely overlooked some edge cases. My latest commit resolves the negative number issue. I’ve looked into several possibilities to address the “-“ or “+” in the middle of the input and did not find a suitable solution. What happens there is when the event is passed from the Input component to the handler, the number is already converted to a string. If the input is non-sensical, such as “100-110” it returns an empty string (“”). This is a behavior of the HTML input tag of type "number", not the Reactstrap component. Because of this behavior, I can’t restrict that specific case without also restricting an empty string entry, which is what happens when a user enters a number and then deletes it. Submitting the form with this input is not going to have harmful effects on the app because the value will be converted to 0. I added a validation for the case when an input is something like “100-110” and the other input (Quantity Wasted, for instance) is zero or “”. In this scenario the “Update” button will be disabled and nothing will happen. In the edge case when one entry is nonsensical and the other is valid, the nonsensical one will be ignored. Given a choice between bad UI and addressing a rare edge case I decided to choose in favor of the UI. If you know a better approach on the front end, please let me know. I also added an extra validation on the back end to send a 200 status when either number is negative. It should never happen if the front end is working correctly but having another level of protection is never a bad idea.
> 
> As for the linting issue, I agree that it could be a little annoying having to scroll through a list of changes that are actually just semicolons and spaces. In theory, previous PR’s with these changes that were merged should have eliminated this problem. I’m not sure why it still exists. If you know the workaround, I suggest you bring it up to Jae, so that he can update the [document on using ESLint and Prettier](https://docs.google.com/document/d/1-zXEU1Go1RheI0lNSY9v8bnvFUjOkp7oJHDT1CxuowM/edit#heading=h.emft21z1gsuy). Right now there is no official policy regarding this.
> 
> @Weiyao-Li and @dhairya-mehta You are right, there was a conflict between my branch and the current version of the “development” branch. I pulled the changes and committed again. Please retest at your convenience. Thank you

Hey Ilya,

I think your approach with the UI. I also tested non sensical inputs and they do not impact negatively the application. The only issue was negative numbers, and if solved, then that's solves the main issue. And, yes, I have talked to Jae about different best practices for coding, formatting, architecteting software, etc, that I will be hopefully helping to bring to both projects (thank you for referrencing your documenation). I will review the changes and approve your PR at this point, since your commit history has considered those edge cases I brought up. Thank you for this exchange.

bhuvan-dv reviewed (dismissed):
The changes are on point, works as expected!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/67942070/74bd100e-e54d-4424-85b6-285ce627e2f3

Coops023 reviewed (dismissed):
Hey Ilya, i have reviewed this PR and all the changes are working as intended! great job!

<img width="1343" alt="Screenshot 2024-03-09 at 11 58 46" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/81972263/e710015e-f8c8-42be-9c28-5e96b0f034b5">

PratimaSingh02 reviewed (dismissed):
Hi, Can we use Yup or Formik for form validation so that no negative values get accepted?
I performed the following steps to review this PR:
1. Logged in as owner
![Screenshot (52)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/48998232/9464cd50-43ef-4563-98d1-3d3e797ca940)
3. Went to /bmdashboard
![Screenshot (53)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/48998232/e93ff57d-a7d6-40ad-a74b-ad086ae0285e)
![Screenshot (54)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/48998232/5255b38c-79e3-400e-9bb5-49f599d8e3d0)

5.  Went to /bmdashboard/consumables
![Screenshot (55)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/48998232/bb805070-8c4e-40a5-976f-e43b0d6b803f)
6. Edited a specific Item
![Screenshot (56)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/48998232/85c2cc16-9cfc-4706-b91b-17746d38b198)
7. Saved the changes
![Screenshot (57)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/48998232/7f3bd0de-ddb1-4c7a-942a-b81cdb70f013)
8. Viewed the changes and they were reflected at the end
![Screenshot (58)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/48998232/c65dd0da-65f6-49d5-ac50-c8db52a8c0c8)

YHX6 reviewed (changes requested):
Hi, I tested the PR review. It works for most of the cases but I run into following issue.
When the quantity is zero and I choose wasted/used in "percent", it will always allow me to do it. Thought it would not affect the database result, but prevent this operation is more user-friendly.
<img width="1624" alt="Screenshot 2024-03-09 at 3 15 17 PM" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113729159/c38f7abf-1363-406d-b555-11f35a557af1">

psharma1984 reviewed (dismissed):
Hi @ilyaflaks The PR works as expected. I was able to update the quantity and see in the updates.

[screen-capture.webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/123996264/42438b94-adb8-48ea-8111-90781df3bd67)

ilyaflaks commented:
> YHX6

Hi, Yaohong. Restricting the user from changing units before a value is entered would not be difficult for me to do, but I'm not sure it's a better UI. A user might want to select the units first and then enter the amount. And if the amount is 0 then the units really don't matter to the user or to the app. This is the UI you can see on all of the forms in the Building Management dashboard and the majority of other applications. 

> PratimaSingh02
Yes, using something like Formik would be a good idea. However, most of the forms for BM Dashboard have already been created and I wouldn't want to complicate things at this point in time and make the app harder to maintain in the future. For now, the negative number issue is resolved. Thank you for your review.

Abi-Liu reviewed (dismissed):
Left a review on the backend PR: https://github.com/OneCommunityGlobal/HGNRest/pull/780

Tiantian-C reviewed (commented):
Hey, I have made a review. It works well. Good job!
<img width="1440" alt="#2010-1" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/122494725/e399d31f-e712-41d0-964b-eb4f6fa8e70c">
<img width="1431" alt="#2020-2" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/122494725/9435bec8-a17f-47b4-ba1c-a3795b3f8d86">
<img width="1435" alt="#2020-3" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/122494725/8484b8d1-74fc-4a2a-b954-ba11b3d0f63e">
<img width="1434" alt="#2020-4" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/122494725/0c477992-d8f2-49b9-9470-a4c9c51ec50b">

brandta-1 reviewed (dismissed):
# Review
Successfully edited building consumables. UI and tables look simple and clear. Tested with Chrome developer tools and MongoDB Compass.

# Pictures

![00_update_consumable](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/116298512/fe75f083-32ed-4268-8520-d673d73e64ac)

![01_updated_table](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/116298512/a838a627-0346-4f4f-a926-942589600d88)

![02_back-end_update](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/116298512/a45929ee-7912-4c58-8498-817225d32db2)

SRamen1999 reviewed (commented):
After testing this PR everything works as expected. I am able to press “View” in the “Updates” column and verify that the change made.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/56932664/eb7e487f-f127-46d2-851d-98a897103009

SanketKaware reviewed (changes requested):
Calculation with default units works for "Used" and "Wasted". If "Available" is zero, I can't click on "Update Consumable" which is correct. 

But when I choose "%" unit, "New Available" field doesn't show up, and I'm able to click "Update Consumable", which is incorrect.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/17793708/ac976eff-0339-4902-9b09-525e44dbda84



Looks like there is an issue with 'validate' function in **UpdateConsumable.jsx**

iven-yao commented:
Hi I just tested out this PR, here's what I've found
1. updating with specified "Units" while "Available" is 0 will be blocked, which is great
2. updating with "%" while "Available" is 0 will still allow user to update consumable, and will create an update record in the bottom of view tab, as expected. updating value will be 0/0, so probably not causing anything, but not sure if this is the logic we need here.
![image](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/25358966/63a84913-975c-41c9-ad99-c9d8285dff33)
![image](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/25358966/4e52140f-6447-4f75-b98d-56016aa7172f)
![image](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/25358966/5ce84ba8-26f9-4356-973b-fc7d3eb71572)
![image](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/25358966/1217127f-16e7-4803-8a1f-c60e0910ef05)

ImzIssa reviewed (changes requested):
Hello, I've tested this PR and I've discovered what @SanketKaware and @iven-yao did.

![Screenshot (35)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/87269725/dee4ba11-55c4-416e-a2a9-2f70ba1fddee)

20chen-7 reviewed (dismissed):
Approved at the backend, good job!

ilyaflaks commented:
@SanketKaware @iven-yao @ImzIssa Thank you for pointing this out! I looks like I did not account for a case when the stock is 0 and any percent of it is automatically zero as well. I have made a very minor change in the "validate" function. Please test again at your convenience. 

@nnamdixobi @Weiyao-Li @dhairya-mehta @Tiantian-C @SRamen1999 Thank you for your reviews. If you find that this PR works well, please change your **comment** to an **approval**. Also, my last commit automatically dismissed all of the old approvals, sorry if yours got dismissed.

hetvi11110 reviewed (dismissed):
I have looked over and tested this PR as it was mentioned. It is working perfectly on my local server. Here is a reference video.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/13557067/09d674b2-ef8a-4f92-8d43-f3d53ce368f0

ImzIssa commented:
> @SanketKaware @iven-yao @ImzIssa Thank you for pointing this out! I looks like I did not account for a case when the stock is 0 and any percent of it is automatically zero as well. I have made a very minor change in the "validate" function. Please test again at your convenience.

Hey @ilyaflaks I've tested again and I've noticed the update button is disabled when the Available is 0 and I use the Quantiyinput fields and the percent unit which is great. 
I noticed too that the New Available is still calculated to a negative value when the stock is 0 and I use the quantity fields, is that deliberate?

![Screenshot (46)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/87269725/588e72e3-754e-4b44-879a-fef13b589d64)

ImzIssa commented:
> > > @SanketKaware @iven-yao @ImzIssa Thank you for pointing this out! I looks like I did not account for a case when the stock is 0 and any percent of it is automatically zero as well. I have made a very minor change in the "validate" function. Please test again at your convenience.
> > 
> > 
> > Hey @ilyaflaks I've tested again and I've noticed the update button is disabled when the Available is 0 and I use the Quantiyinput fields and the percent unit which is great. I noticed too that the New Available is still calculated to a negative value when the stock is 0 and I use the quantity fields, is that deliberate?
> 
> Yes! That is deliberate. I want to keep some way of alerting the user that they are trying to update the stock to a negative value and 5% of zero is still zero, which is not negative. Plus it triggers validation warnings the same way as with units.

@ilyaflaks okay thanks. I've approved your changes.

SanketKaware commented:
> @SanketKaware @iven-yao @ImzIssa Thank you for pointing this out! I looks like I did not account for a case when the stock is 0 and any percent of it is automatically zero as well. I have made a very minor change in the "validate" function. Please test again at your convenience.
> 
@ilyaflaks I rechecked it. Now it's working as expected.

Approved both frontend and backend PR [780](https://github.com/OneCommunityGlobal/HGNRest/pull/780).

Here is my today's review video for reference - 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/17793708/7b39a054-258a-4d0b-a3b0-23efee287bb5

Gayathr-2012 commented:
HI Ilya,

logged  in as any user
 went to http://localhost:3000/bmdashboard and log in again
went to http://localhost:3000/bmdashboard/consumables, select any item on the table, press the “pencil” icon, enter some numbers next to “Quantity Used” and/or “Quantity Wasted”. Select "%" as units or leave the item's default units 
Press “Update Consumable”
A success toast message should appear if everything is acceptable or a warning message if not
From the Consumables table, press “View” in the “Updates” column and verify that the change you made is reflected. The latest update should be on the bottom

What you have mentioned I have followed these steps It works as expected.

<img width="1255" alt="image" src="https://github.com/OneCommunityGlobal/HGNRest/assets/163784958/886dbfcf-7a19-4067-a3f6-7660a8ba8c88">

<img width="1241" alt="image" src="https://github.com/OneCommunityGlobal/HGNRest/assets/163784958/b05908c4-bef7-4164-8e6c-2d6bc1a5aba8">

<img width="928" alt="image" src="https://github.com/OneCommunityGlobal/HGNRest/assets/163784958/47f7bb84-3915-4e9f-a4bb-16d52e7f0f31">

<img width="1098" alt="image" src="https://github.com/OneCommunityGlobal/HGNRest/assets/163784958/0f99cb26-bda3-4f12-b7fc-6a47821ea473">

<img width="924" alt="image" src="https://github.com/OneCommunityGlobal/HGNRest/assets/163784958/81288466-2a44-4f95-aa1b-94051b5010b6">

kylepham reviewed (approved):
I have checked the validations and it worked as expected. New values and toast are shown correctly.

<img width="1800" alt="Screenshot 2024-03-27 at 4 09 41 PM" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/35129229/cd1eae59-86b7-415f-bf01-a9bfcb570888">

![Screenshot 2024-03-27 at 4 09 41 PM (2)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/35129229/4f88ea08-6ec4-4811-86c7-90781e722693)

![Screenshot 2024-03-27 at 4 09 00 PM (2)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/35129229/e6cf0698-7ebe-43fc-9bdd-be8bbb404c8f)

![Screenshot 2024-03-27 at 4 07 00 PM (2)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/35129229/0158dc77-f61e-451c-a587-07e30050aa43)

kylepham commented on src/components/BMDashboard/UpdateConsumable/UpdateConsumable.jsx:
```diff
@@ -0,0 +1,310 @@
+    );
+  }, [updateRecord]);
+
+  const validate = (_qtyUsed, _qtyWasted, QtyUsedLogUnit, QtyWastedLogUnit) => {
```
These variables `QtyUsedLogUnit`, `QtyWastedLogUnit` doesn't seem to conform with common naming convention in JS.

ilyaflaks replied:
@kylepham  you have a point. These variable names mostly carried over from similar components created by other people. I made some changes to resolve merge conflicts and also changed these variable names on the front end and the backend. Please review again when you have a chance and give me your approval if everything works

Jaiwin-4998 reviewed (dismissed):
Hi Ilya, I just check you PR and it is working fine as intended.
<img width="1440" alt="PR #2010" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/62836815/df4315e6-5675-4a1d-b913-e272d960fa35">
<img width="1440" alt="PR 2010" src="https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/62836815/bc476695-780e-49a7-8938-025719cee585">

ilyaflaks committed:
fix: adjusted to fit new component structure

ilyaflaks committed:
Merge branch 'development' into ilya_consumables_single_update
resolving merge conflicts stemming from other commits merged with development.

metaphor987 reviewed (dismissed):
Hi Ilya,

I've tested the main functionalities using owner, admin and volunteer accounts and they all work well. The "Update Form" handles input validation, warning users about out-of-range inputs, which helps prevent errors. Nice work!

Here are some small suggestions for further improvement:
- It would be better to prevent users from typing in "-" or "+" in the input fields, or show some warnings in these cases, although currently inputs containing these characters will not lead to actual changes in the data.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/91745551/22924533-a1de-4c3a-af6e-907618901666


- Displaying numbers in the "Updates" -> "Views" tab with about four decimal places may have a better look.
![Screenshot 2024-04-17 at 3 33 25 PM](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/91745551/bc8fe31b-a842-4b48-8e00-de542357331b)

Thank you!

jennZhang93 reviewed (dismissed):
Tested with owner account and I was able to edit the Quantity Used and Quantity Wasted field. The success message also pop up. And my update history showed up correctly as well.

![Screenshot 2024-04-17 at 9 34 46 PM (2)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/66922622/016e6d07-39d3-4a45-8083-0acd428bf5a1)
![Screenshot 2024-04-17 at 9 34 50 PM (2)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/66922622/32ccf4c2-185f-401e-bae2-73c5452127f4)
![Screenshot 2024-04-17 at 9 35 20 PM (2)](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/66922622/b4587168-e80b-451f-bd81-0b36dd305954)

aaryaneil reviewed (dismissed):
Tested the PR and the changes work as mentioned. Good work.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/39597927/c478dacb-83df-464a-8948-7d54f9f5b135

ilyaflaks committed:
merge conflicts

ilyaflaks committed:
Merge branch 'development' into ilya_consumables_single_update

ilyaflaks committed:
fix: merge conflicts

ilyaflaks committed:
fix: lint errors

ilyaflaks commented:
@one-community Please merge this branch into "development". All of the requested changes have been addressed

one-community commented:
Thank you all, merging!
